### PR TITLE
KAFKA-15890: Consumer.poll with long timeout unaware of assigned partitions

### PR DESCRIPTION
**Pull request by AndrewJSchofield**
In the new consumer, `Consumer.poll(Duration timeout)` blocks for the entire duration. If the consumer is joining a group and has not yet received its assignments, the poll begins before an assignment has yet been received. Because the poll is blocked, it does not notice when partitions are assigned, and it subsequently does not return any records. The old consumer only blocks for the duration of the heartbeat interval and loops for until the poll timeout has passed, and is thus able to check for assignments received.

When this problem has been fixed, there remains another which prevents the group becoming stable. Because the consumer repeatedly sends the list of topic-partitions that it has been assigned to the group coordinator, the coordinator responds with the list of topic-partitions, which causes the consumer to remain reconciling indefinitely. By making the building of `ConsumerGroupHeartbeatRequest` stateful, the loop is ended and the group becomes stable as expected.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: we should check for empty string (`memberId` defaults to empty string and is then updated when received from the coordinator)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: typo successfully

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1077]: Nice, this was definitely missing. Just wonder if it would be better to move the transition to reconciling to the reconcile function itself, to have it in a single place.

Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1077]: I think I'll leave this part up to you in the next revision.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: I see what you mean. Actually, I think no check is needed. Initially, it's empty "" which causes the coordinator to create a UUID for the member ID. Subsequently, it's necessary to send the member ID on every request. I've just removed the check.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:379]: This is perfectly fine, but just for the record, it is not strictly needed. When the subscription changes (updated in the subscription state), it is picked up and sent out on the next heartbeat request, without requiring a transition to joining.

**Comment by lianetm** [comment]
Thanks for the changes @AndrewJSchofield! LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Nit: leaving the indentation as is conforms to the existing 'four-spaces per tab' style.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:203]: Nit: we got knocked in other reviews for unnecessarily qualifying with `this`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:400]: Any reason `HeartbeatState` needs to be `static`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: IMO, we should group all of the 'last sent' attributes in a separate little struct/class, just to keep their relation and their (re-)initializing logic together.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]:
```suggestion
                partitionsPerTopicId.computeIfAbsent(topicId, __ -> new ArrayList<>()).add(partitionId);
```

**Review by kirktrue — APPROVED** [review]
Thanks for the PR, @AndrewJSchofield 😄 I have some minor nits, but I'll approve this and leave it up to you to change anything.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:400]: So that you can construct an independent instance of the class. It doesn't need a `this` and it's handy for testing to be able to make instances on demand.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:203]: That style is used throughout this file.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:400]: Also, spotbugs will object if an inner class is non-static and does not use its `this` reference to the outer class.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Done

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Looks like the subscription field is now unused.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: `.collect(Collectors.toCollection(TreeSet::new)` and skip the intermediate list. 

Also, you seem to avoid using `TopicIdPartition.toString` or just `TopicIdPartition.equals` here - is it to support topic renaming? If so, I think it's quite subtle and would deserve a little comment.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: I think in this comment and the similar ones below, you wanted to say `send if it _has_ changed`

**Review by lucasbru — COMMENTED** [review]
Thanks for the PR! prod code looks good to me, I just added three minor things that I spotted while going through the code. However, I have two requests:

 - Can we add a test to check that we are interrupting the wait correctly in poll? I could imagine if we mock `timeUntilNextPoll` it shouldn't be too difficult.

 - It seems we are solving two things in this PR - correctly build the heartbeat (only send the diff) and the fix to not miss something. We don't need to split the PR, but could you update the PR description?

**Comment by AndrewJSchofield** [comment]
> Thanks for the PR! prod code looks good to me, I just added three minor things that I spotted while going through the code. However, I have two requests:

> Can we add a test to check that we are interrupting the wait correctly in poll? I could imagine if we mock timeUntilNextPoll it shouldn't be too difficult.

Certainly. I'll do it.

> It seems we are solving two things in this PR - correctly build the heartbeat (only send the diff) and the fix to not miss something. We don't need to split the PR, but could you update the PR description?

Yes, I'll do that. We just needed both things to make the `kafka-console-consumer.sh` tool correctly join a stable group and start receiving records.

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!

**Comment by lucasbru** [comment]
@AndrewJSchofield 

Looks like the test is failing 

```
org.mockito.exceptions.misusing.WrongTypeOfReturnValue: 
Long cannot be returned by poll()
poll() should return PollResult
***
If you're unsure why you're getting above error read on.
Due to the nature of the syntax above problem might occur because:
1. This exception *might* occur in wrongly written multi-threaded tests.
   Please refer to Mockito FAQ on limitations of concurrency testing.
2. A spy is stubbed using when(spy.foo()).then() syntax. It is safer to stub spies - 
   - with doReturn|Throw() family of methods. More in javadocs for Mockito.spy() method.
```

Could you please have a look?

**Comment by AndrewJSchofield** [comment]
Looks like you got there slightly earlier. I also noticed that the same test failure needs addressing. I'll investigate.

**Comment by lucasbru** [comment]
@AndrewJSchofield looks like there are related test failures

**Comment by AndrewJSchofield** [comment]
@lucasbru I agree. The change has made one or two tests flaky. I will fix this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:227]: This will invoke the request managers directly from the application thread, right?

If so, I'm a little concerned by this approach because the implementation of `maximumTimeToWait` in each `RequestManager` will be executed by the application thread. In the implementations elsewhere in this PR we're reading and writing state in the request managers that is only intended to be read or written by the network I/O thread.

We need to be careful here 🤔

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:818]: Example, this is updating the Timer from both threads.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java:400]: Fair enough 😄

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java:1077]: OK, I think I'm running into this too. In my tests, the state machine was preemptively `RECONCILING` only to later find out the assignments were the same, after which it returned from `reconcile()` without resetting the state.

**Review by kirktrue — CHANGES_REQUESTED** [review]
@AndrewJSchofield I took another look at this and there's potentially a problem that we'd need to discuss. I left some comments, but we should clarify before merging this.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:227]: Yes, we do. In the previous commit, I inadvertently ended up with a ConcurrentModificationException for this very reason. This is why I scaled back the change to be only accessed the timer.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:818]: True. I think it's an atomic operation which is updating the timer to a more current value.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:227]: I'm wondering how "fresh" the "maximum time to wait" value needs to be...

Spitballing here, but... say we introduce a new instance variable in `ConsumerNetworkThread` that stores a cached value named, say, `notTotallyFreshMaxWaitCached`? We could then invoke `maximumTimeToWait()` from inside `runOnce()` and update `notTotallyFreshMaxWaitCached`, like so:

```java
    private volatile long notTotallyFreshMaxWaitCached = MAX_POLL_TIMEOUT_MS;

    void runOnce() {
        // If there are errors processing any events, the error will be thrown immediately. This will have
        // the effect of closing the background thread.
        applicationEventProcessor.process();

        final long currentTimeMs = time.milliseconds();
        final long pollWaitTimeMs = requestManagers.entries().stream()
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(rm -> rm.poll(currentTimeMs))
                .map(networkClientDelegate::addAll)
                .reduce(MAX_POLL_TIMEOUT_MS, Math::min);
        networkClientDelegate.poll(pollWaitTimeMs, currentTimeMs);

        notTotallyFreshMaxWaitCached = requestManagers.entries().stream()
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(rm -> rm.maximumTimeToWait(currentTimeMs))
                .reduce(Long.MAX_VALUE, Math::min);
    }

    public long maximumTimeToWait() {
        return notTotallyFreshMaxWaitCached;
    }
```

(Or something like that.)

This way we still only touch the `RequestManager`s inside `runOnce()`, so we don't open overselves up to subtle concurrency issues.

Thoughts?

Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:227]: Something like this would work I think. The aim after all is to keep things responsive rather than an accurate regular tick.

Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:227]: I'll take this concept and work on it in a follow-on PR.

**Review by lucasbru — APPROVED** [review]

**Pull request merged**
